### DISCUSSION TITLE
PrimeVue Dynamic Dialog Slots

### DISCUSSION THREAD

**Discussion by op-jsn** [primefaces/community#1208] [PrimeVue]
Hey I'm currently looking into replacing the [Dialogs](https://primevue.org/dialog/) in my application to [DynamicDialogs](https://primevue.org/dynamicdialog/).

I'm currently stuck trying to override the default Close & Maximize icons in DynamicDialogs as I have done with standard Dialogs.

A similar question was asked [here](https://github.com/orgs/primefaces/discussions/185) but doesn't go into detail for any other slots.

As per [the Dynamic Dialog documentation](https://primevue.org/dynamicdialog/#api.dialog.slots), `closeicon` and `maximizeicon` slots should be available.

However, the exported `DynamicDialogTemplates` interface does not include these slots:

`
export interface DynamicDialogTemplates {
   /**
    * Custom header template.
    */
   header?: any;
   /**
    * Custom footer template.
    */
   footer?: any;
}
`

Additionally, the slots are not available within the template. This is how I did it with standard Dialogs.

`<DynamicDialog>
  <template #maximizeicon>
     //my new icon here//
  </template>
</DynamicDialog>
`

Just wondering if I am doing something wrong here or if the documentation is incorrect?

Thanks for any help.

**Comment by digitalextremist** [comment]
Was directed here from `discord` when I noticed the same issue and asked about it.

Not only is this a known issue, it is also on my radar to be my first contribution here... ever/soon.

I _suspect_ this is related to the other issue I asked about, which is the revealing of the `maximize` state, and ability to trigger `maximize` dynamically. That is also not available/accessible to the component, which the icon runs alongside. It's a stated icon.

Alongside `#maximizeicon`, yes I would prioritize `#header` too for `DynamicDialog` to be possible to swap in for `Dialog` across the board, which it currently cannot do. Might take a while to get to, but that is the motivation for me to get this resolved too.

**Comment by kreesty** [comment]
I'd also like for this to be possible. Dynamic Dialogues are much cleaner than regular ones, but they need to support customization.
In my case, I'd like to be able to modify the header and show an image to the left of the title.

**Reply by digitalextremist** [reply]
Same page, exactly the same requirements it seems. Am feeling the pain! You feel a PR coming on @kreesty? 😆

**Reply by kreesty** [reply]
> Same page, exactly the same requirements it seems. Am feeling the pain! You feel a PR coming on @kreesty? 😆

Turns out there's a way to do it! There's a 'templates' struct available inside the DynamicDialogOptions.
Here's what worked for me on PrimeVue 4.3.9:

```
// build header VNode with thumbnail on the left
const imgSrc = productImageMap.value[selectedProduct.value.id!];
    const headerVNode = h('div', { style: 'display:flex;align-items:center;gap:.6rem;' }, [
        imgSrc
            ? h('img', {
                src: imgSrc,
                alt: 'Product',
                style: 'width:36px;height:36px;object-fit:cover;border-radius:6px;flex:0 0 36px;'
            })
            : h('div', {
                style: 'width:36px;height:36px;display:flex;align-items:center;justify-content:center;background:var(--p-surface-b);border-radius:6px;flex:0 0 36px;'
            }, [h('i', { class: 'pi pi-box', style: 'font-size:1.1rem;color:var(--p-text-color-secondary);' })]),
        h('span', { style: 'font-weight: var(--p-dialog-title-font-weight); font-size: var(--p-dialog-title-font-size);' }, `${selectedProduct.value.name}`)
    ]);

    dialog.open(productDetailsComponent, {
        templates: {
            header: headerVNode
        },
        props: { ... },
        data: { ... },
        emits: { ... }
    });
```

You can also do something similar for the footer.

**Reply by digitalextremist** [reply]
Nice find @kreesty! Looks scary to do in-the-raw, but looks like it might be a hold-over until it can be done in `<template>`!
